rotkehlchen/tests/exchanges/test_kucoin.py

```diff
@@ -0,0 +1,727 @@
+import warnings as test_warnings
+from contextlib import ExitStack
+from http import HTTPStatus
+from json.decoder import JSONDecodeError
+from unittest.mock import patch
+
+import pytest
+import requests
+
+from rotkehlchen.accounting.structures import Balance
+from rotkehlchen.assets.asset import Asset
+from rotkehlchen.assets.converters import UNSUPPORTED_KUCOIN_ASSETS, asset_from_kucoin
+from rotkehlchen.errors import RemoteError, UnknownAsset, UnsupportedAsset
+from rotkehlchen.exchanges.data_structures import AssetMovement, Trade, TradeType
+from rotkehlchen.exchanges.kucoin import Kucoin, KucoinCase, SkipReason
+from rotkehlchen.fval import FVal
+from rotkehlchen.tests.utils.mock import MockResponse
+from rotkehlchen.typing import (
+    AssetAmount,
+    AssetMovementCategory,
+    Fee,
+    Location,
+    Price,
+    Timestamp,
+    TradePair,
+)
+from rotkehlchen.utils.serialization import rlk_jsonloads_dict
+
+
+def test_name():
+    exchange = Kucoin(
+        api_key='a',
+        secret=b'a',
+        database=object(),
+        msg_aggregator=object(),
+        passphrase='a',
+    )
+    assert exchange.name == str(Location.KUCOIN)
+
+
+def test_kucoin_exchange_assets_are_known(mock_kucoin):
+    request_url = f'{mock_kucoin.base_uri}/api/v1/currencies'
+    try:
+        response = requests.get(request_url)
+    except requests.exceptions.RequestException as e:
+        raise RemoteError(
+            f'Kucoin get request at {request_url} connection error: {str(e)}.',
+        ) from e
+
+    if response.status_code != HTTPStatus.OK:
+        raise RemoteError(
+            f'Kucoin query responded with error status code: {response.status_code} '
+            f'and text: {response.text}',
+        )
+    try:
+        response_dict = rlk_jsonloads_dict(response.text)
+    except JSONDecodeError as e:
+        raise RemoteError(f'Kucoin returned invalid JSON response: {response.text}') from e
+
+    # Extract the unique symbols from the exchange pairs
+    unsupported_assets = set(UNSUPPORTED_KUCOIN_ASSETS)
+    for entry in response_dict['data']:
+        symbol = entry['currency']
+        try:
+            asset_from_kucoin(symbol)
+        except UnsupportedAsset:
+            assert symbol in unsupported_assets
+        except UnknownAsset as e:
+            test_warnings.warn(UserWarning(
+                f'Found unknown asset {e.asset_name} in kucoin. '
+                f'Support for it has to be added',
+            ))
+
+
+def test_api_query_retries_request(mock_kucoin):
+
+    def get_response():
+        results = [
+            """{"code":400007,"msg":"unknown error"}""",
```

Comment:
why is it the same code/msg response repeated 2 times here?

Reply:
It really does not matter here as I'm using the response status code but not the error code. They didn't answer me if there was a more specific status code for the request limits, and the other libraries (e.g. cctx) don't use the error code neither. I reckon it is 400007 but I can't be sure. I've put two responses as I've limited the retries to 1 in this test.